### PR TITLE
[FLINK-25995][table-planner] Make implicit assumption of SQL local hash explicit

### PR DESCRIPTION
**Pull request by godfreyhe**
## What is the purpose of the change

*Make implicit assumption of SQL local hash explicit: Find the implicit hash distribution by ForwardHashExchangeProcessor, and add a special Exchange to generate ForwardForConsecutiveHashPartitioner.*

## Brief change log

  - *give correct `requiredDistribution` for all batch physical nodes where require hash distribution*
  - *Add ForwardHashExchangeProcessor to add a special Exchange to generate ForwardForConsecutiveHashPartitioner*

## Verifying this change

This change added tests and can be verified as follows:
  - *Added ForwardHashExchangeTest to verify the plans*
  - *Added ForwardHashExchangeITCase to verify the execution result*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Comment by godfreyhe** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/ForwardHashExchangeProcessor.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+                        if (changed) {
+                            node.setInputEdges(newEdges);
+                        }
+                    }
+                };
+        execGraph.getRootNodes().forEach(s -> s.accept(visitor));
+        return execGraph;
+    }
+
+    private boolean hasExchangeInput(ExecEdge edge) {
```
Comment (wenlong88) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/ForwardHashExchangeProcessor.java:128]: rename to isExchangeInput

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/ForwardHashExchangeProcessor.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                                                .damBehavior(inputProperty.getDamBehavior())
+                                                .priority(inputProperty.getPriority())
+                                                .build();
+                                BatchExecExchange newExchange =
+                                        new BatchExecExchange(
+                                                newInputProperty,
+                                                (RowType) edge.getOutputType(),
+                                                newInputProperty.toString());
+
+                                ExecEdge newEdge1 =
```
Comment (wenlong88) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/ForwardHashExchangeProcessor.java:98]: a suggestion: how about add an updateInput(ExecNodeBase newInput, int inputIndex) to ExecNodeBase, and BatchExecMultiInput rewrite the method, and update the input of the original node. 
we could make the magic of MultiInput internal in BatchMultiInput and make it easier to understand the logic here.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/ForwardHashExchangeProcessor.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+ * A {@link ExecNodeGraphProcessor} which finds all {@link ExecNode}s that require hash
+ * distribution, but its input has no hash Exchange node. Currently, the hash distribution is
+ * satisfied by its none exchange inputs with FORWARD partitioner (the node and its input has the
+ * same parallelism). Once the parallelism is changed, the FORWARD behavior will be broken, and the
+ * result will be wrong.
+ *
+ * <p>In order to meet the needs of flexible parallelism changing, a special {@link
+ * BatchExecExchange} (with KEEP_INPUT_AS_IS distribution flag) will be added for the {@link
+ * ExecNode} as its input. And then the runtime will decide which partitioner can be used when
+ * adaptive scheduler is enabled: FORWARD partitioner if the parallelism is the same, or HASH
```
Comment (wenlong88) [inline flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/ForwardHashExchangeProcessor.java:49]: in currently implementation of runtime, the partitioner is chosen at compile time. when it is not chained, the partitioner would be hash ignoring the parallelism.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecExchange.java [hunk 4]
```diff
@@ -126,22 +130,25 @@ public String getDescription() {
...
-                                        keys),
-                                fieldNames);
+                        createHashPartitioner(
+                                ((HashDistribution) requiredDistribution), inputType, planner);
                 parallelism = ExecutionConfig.PARALLELISM_DEFAULT;
                 break;
+            case KEEP_INPUT_AS_IS:
+                RequiredDistribution inputDistribution =
+                        ((KeepInputAsIsDistribution) requiredDistribution).getInputDistribution();
+                checkArgument(
```
[the remark this answers is not in the document]
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/batch/BatchExecExchange.java:141]: for singleton distribution, add the (max) parallelism limitation in ExecNodeBase#translateToPlan

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/ForwardHashExchangeProcessor.java (hunk 2 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/ForwardHashExchangeProcessor.java:98]: It's better we keep the logic in one class

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/ForwardHashExchangeProcessor.java (hunk 3 above)
Comment (godfreyhe) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/ForwardHashExchangeProcessor.java]: updated

**Comment by wenlong88** [comment]
LGTM

**Comment by godfreyhe** [comment]
Thanks for the review, I will merge it

**Pull request closed without merging**